src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java

```diff
@@ -237,7 +236,21 @@ public boolean isForceExecution() {
         @Override
         public void messageReceived(final ReplicaOperationRequest request, final TransportChannel channel) throws Exception {
             try {
-                shardOperationOnReplica(request);
+                IndexShard indexShard = indicesService.indexServiceSafe(request.shardId.getIndex()).shardSafe(request.shardId.id());
+                boolean counterInc = false;
+                try {
+                    counterInc = indexShard.beforeWriteOperation();
```

Comment:
this should be out of the try finally. and then we can avoid the counterInc boolean.